src/test/java/hudson/plugins/git/GitSCMTest.java

```diff
@@ -290,6 +290,137 @@ public void testSpecificRefspecs() throws Exception {
+    }
+
+    /**
+     * After avoiding the second fetch call in retrieveChanges(), this test verifies there is no data loss by fetching a
+     * repository(git init + git fetch) with a narrow refspec with CloneOption of honorRefspec = true on initial clone
+     * First fetch -> narrow refspec (since refspec is honored on initial clone)
+     * Second fetch -> narrow refspec (avoided)
+     **/
+    @Test
+    @Issue("JENKINS-56404")
+    public void testAvoidRedundantFetchWithHonorRefSpec() throws Exception {
+        List<UserRemoteConfig> repos = new ArrayList<>();
+        String refSpec = "+refs/heads/foo:refs/remotes/foo";
+        repos.add(new UserRemoteConfig(testRepo.gitDir.getAbsolutePath(), "origin", refSpec, null));
+
+        /* With honor refspec on initial clone */
+        FreeStyleProject projectWithMaster = setupProject(repos, Collections.singletonList(new BranchSpec("master")), null, false, null);
+        CloneOption cloneOptionMaster = new CloneOption(false, null, null);
+        cloneOptionMaster.setHonorRefspec(true);
+        ((GitSCM)projectWithMaster.getScm()).getExtensions().add(cloneOptionMaster);
+
+        // create initial commit
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit in master");
+        // Add another branch 'foo'
+        git.branch("foo");
+        git.checkout().branch("foo");
+        commit(commitFile1, johnDoe, "Commit in foo");
+
+        // Build will be failure because the initial clone regards refspec and fetches branch 'foo' only.
+        FreeStyleBuild build = build(projectWithMaster, Result.FAILURE);
+
+        FilePath childFile = returnFile(build);
+        if (childFile != null) {
+            // assert that no data is lost by avoidance of second fetch
+            assertThat(childFile.readToString(), not(containsString("master")));
+            assertThat("foo branch was not fetched", childFile.readToString(), containsString("foo"));
+        }
+        assertRedundantFetchIsTrue(build, refSpec);
+
+        assertThat("FAILURE", is((build.getResult().toString())));
+    }
+
+    // Checks if the second fetch is being avoided
+    private void assertRedundantFetchIsTrue(FreeStyleBuild build, String refSpec) throws IOException {
+        List<String> values = build.getLog(Integer.MAX_VALUE);
+        int countFetches = 0;
+        String argRefSpec = " " + refSpec;
+        String fetchArg = " > git fetch --tags --force --progress -- " + testRepo.gitDir.getAbsolutePath() + argRefSpec + " # timeout=10";
+        for (String value : values) {
+            if(value.equals(fetchArg)){
+                countFetches++;
+            }
+        }
+        // Before the flag check, countFetches was "2" because git fetch was called twice
+        assertThat(countFetches, is(not(2)));
+
+        // After the fix, git fetch is called exactly once
+        assertThat(countFetches, is(1));
```

Comment:
Redundant check.
```suggestion
        // Before the flag check, countFetches was "2" because git fetch was called twice
        // After the fix, git fetch is called exactly once
        assertThat(countFetches, is(1));
```

Reply:
This is more of a descriptive statement than a check, what is meant to be shown is that after the fix, the number of fetches are reduced from "2" to "1".